### DISCUSSION TITLE
cfde-submit validation error from sample substance.tsv

### DISCUSSION THREAD

**Discussion by jeremyjyang** [nih-cfde/published-documentation#330] [Q&A] [answered]
cfde-submit -ing (--dry-run for validation) a datapackage today with sample substance.tsv, I get this error:
`Validation error in substance.tsv
Type error in the cell "['benzaldehyde', '100-52-7', 'benzanoaldehyde']" in row "2" and field "synonyms" at position "4": type is "array/default"`
Reading from https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-substance.tsv
should the example row be deleted before submission?

**Comment by ACharbonneau** [comment]
@abradyIGS do you know what is causing this?

**Comment by abradyIGS** [comment]
Typo in the wiki; apologies: the synonyms column was missing. Fixed now, double-checking all other table pages now.

**Comment by abradyIGS** [comment]
Update: all wiki pages now triple checked: this was the only missing field. Have updated & clarified a bunch of the table description blurbs on various pages in addition to field-name checks. Please yell if anything at all still seems amiss.

**Reply by jeremyjyang** [reply]
Still getting this error:

> [DEBUG] cfde_submit.main::run() Loaded previous state
> [DEBUG] cfde_submit.main::run() Determining DCC
> [DEBUG] cfde_submit.main::run() Initializing Flow
> [INFO] cfde_submit.client::check() Check PASSED, client is ready use flows.
> [DEBUG] cfde_submit.main::run() CfdeClient initialized, starting Flow
> Submit datapackage 'submission' using cfde_registry_dcc:idg? (y/n)? > y
> [INFO] cfde_submit.client::check() Check PASSED, client is ready use flows.
> [DEBUG] cfde_submit.client::start_deriva_flow() Startup: Validating input
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Checking for a Git repository
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Not a Git repo
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Creating BDBag out of directory '/home/data/CFDE/data/drugpages2021/submission'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Copying data to '/home/data/CFDE/data/drugpages2021/submission_output' before creating BDBag
> [DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag created at '/home/data/CFDE/data/drugpages2021/submission_output'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Archiving BDBag at '/home/data/CFDE/data/drugpages2021/submission_output' using 'zip'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag archived to file '/home/data/CFDE/data/drugpages2021/submission_output.zip'
> [DEBUG] cfde_submit.validation::validate_user_submission() Validating TableSchema in BDBag '/home/data/CFDE/data/drugpages2021/submission_output.zip'
> Validation error in substance.tsv
> Type error in the cell "['benzaldehyde', '100-52-7', 'benzanoaldehyde']" in row "2" and field "synonyms" at position "4": type is "array/default"

**Comment by abradyIGS** [comment]
Please attach offending substance.tsv? First five rows should suffice if it's big

**Reply by jeremyjyang** [reply]
It appears the offending substance.tsv is being generated by prepare_C2M2_submission.py. Two data rows only.

> id	name	description	synonyms	compound
> 2161	benzaldehyde		['benzaldehyde', '100-52-7', 'benzanoaldehyde']	240
> 2163	benzoate		['benzoate', '65-85-0', 'benzoic acid', 'benzenecarboxylic acid', 'phenylformic acid']	242

**Comment by abradyIGS** [comment]
Assuming there are two tab characters between the values in the `name` and `synonyms` fields, that's compliant -- @jeremyjyang please confirm? (i can't see tabs in the inline display)

@karlcz do you happen to recognize what's causing the error reported two comments up?

@jeremyjyang it might help streamline things if you could double-check that the C2M2_datapackage.json you're submitting with is an exact match for the one [here](https://osf.io/vzgx9/). (TBC this doesn't seem to me like an issue with your local schema copy, but as long as i don't understand what's going on, i figure standard checks can't hurt)

**Reply by ACharbonneau** [reply]
@karlcz could you double check that the submission system is set to expect an array for the synonyms column for substance? It looks like the error from yesterday is a type error in the submission code to me.

**Reply by jeremyjyang** [reply]
Sorry for slow response, but yes to question from @abradyIGS :

> $ wc -l /home/data/CFDE/data/diseasepages6124/submission/C2M2_datapackage.json ~/Downloads/C2M2_datapackage.json 
>   2820 /home/data/CFDE/data/diseasepages6124/submission/C2M2_datapackage.json
>   2820 /home/jjyang/Downloads/C2M2_datapackage.json
>   5640 total
> $ diff /home/data/CFDE/data/diseasepages6124/submission/C2M2_datapackage.json ~/Downloads/C2M2_datapackage.json

**Comment by karlcz** [comment]
I think that error is coming from frictionless-py, not from our CFDE-specific code.  If the error message is showing the actual value (not transforming it in some way), then the offending array is not being supplied in JSON format.  The meta-syntax for the synonyms array of strings must follow JSON:

`[`  `"` _string content_ `"` `,` ... `]`

i.e. each string is surrounded by the double-quote character `"`, not the single quote/apostrophe `'` which I see in the error message. 

Please provide a URL to the exact input files if you want me to continue to investigate this...

**Comment by abradyIGS** [comment]
This is a problem with my preprocessing code and is also related to [this issue](https://github.com/nih-cfde/published-documentation/discussions/337) -- will push a fix ASAP, need a day or so to reprocess master files

**Comment by abradyIGS** [comment]
Problem has been identified and fixed; I'm doing final checks on my reprocessing pipeline now and I expect to upload new PubChem reference tables to OSF today that will solve this and a couple of other reported issues.

(note to self: stop using phrases like "a day or so")

**Reply by jeremyjyang** [reply]
Please report status and indicate when the fix should be available, and whether via reference tables, and/or CV files, and/or revised submission script, etc. I see there are new CV files dated April 1.

**Comment by abradyIGS** [comment]
Status is imminent; today is public release day for the new features. Script, final CV files and explicit responses to all pending issues will be delivered by (or shortly after) COB Eastern. Apologies for delays: our schedule was accelerated this quarter by 50%. Thank you for your patience and tire-kicking during pre-deployment testing.

**Comment by abradyIGS** [comment, marked as answer]
This issue has now tested as "fixed" -- marking "answered," but please reopen and report if you find otherwise.

Ontology files and the submission prep script have all been updated and published [here](https://osf.io/bq6k9/): please overwrite local copies and try again.

Note: the master C2M2 JSON schema has also updated (to match previous development schema): it is [here](https://osf.io/c63aw/) and an up-to-date copy needs to be included with submissions during prep with the script.

**Reply by jeremyjyang** [reply]
Great, thanks, much better, almost there it seems. One thing I notice the sample collection.tsv is missing column has_time_series_data. I should probably be using the schema only and not the sample TSVs I suppose.

> Validation error in collection.tsv
> There is a missing label in the header's field "has_time_series_data" at position "8"

**Reply by abradyIGS** [reply]
Thanks for catch - have pushed a fix to the sample TSVs on OSF.
